### PR TITLE
e2e: Migrate duplicate node id test from kurtosis

### PR DESCRIPTION
**Pull request by marun**
## Why this should be merged

Fulfills one of the requirements for #1547 (migration of Kurtosis tests)

## How this works

- adds node add/stop to testnet fixture
- reimplements avalanche testing's [duplicate node test](https://github.com/ava-labs/avalanche-testing/blob/master/tests/duplicate_node_id.go)

## How this was tested

Ran ./scripts/test.e2e.sh

## TODO

- [x] Merge the parent PR (#1709)

### CODE REVIEW

File: tests/e2e/e2e.go [hunk 1]
```diff
@@ -26,6 +29,10 @@ const (
 	// Enough for test/custom networks.
 	DefaultConfirmTxTimeout = 20 * time.Second
 
+	// This interval should represent the upper bound of the time
+	// required to start a new node on a local test network.
+	DefaultNodeStartTimeout = 20 * time.Second
```
Comment (hexfusion) [inline tests/e2e/e2e.go:34]: Just curious what the basis for this timeout is? Do we expose metrics on this?

File: tests/e2e/e2e.go [hunk 2]
```diff
@@ -74,3 +81,32 @@ func (te *TestEnvironment) AllocateFundedKeys(count int) []*secp256k1.PrivateKey
...
+	// A temporary location ensures the node won't be accessible from
+	// the Network instance.
+	flags[config.DataDirKey] = ginkgo.GinkgoT().TempDir()
+
+	node, err := network.AddNode(ginkgo.GinkgoWriter, flags)
+	require.NoError(err)
+
+	// Ensure node is stopped and its configuration removed on teardown
+	ginkgo.DeferCleanup(func() {
+		tests.Outf("Shutting down temporary node %s\n", node.GetID())
```
Comment (hexfusion) [inline tests/e2e/e2e.go:100]:
```suggestion
		tests.Outf("Shutting down temporary node: %s\n", node.GetID())
```

File: scripts/tests.e2e.sh [hunk 3]
```diff
@@ -43,10 +43,27 @@ else
...
-#   The test binary by itself isn't capable of running specs in parallel.
+# Determine ginkgo args
+GINKGO_ARGS=""
+if [[ -n "${E2E_SERIAL:-}" ]]; then
+  # Specs will be executed serially. This supports running e2e tests in CI
+  # where parallel execution of tests that start new nodes beyond the
+  # initial set of validators could overload the free tier CI workers.
+  # Forcing serial execution in this test script instead of marking
+  # resource-hungry tests as serial supports executing the test suite faster
+  # on powerful development workstations.
```
Comment (hexfusion) [inline scripts/tests.e2e.sh:54]: Might be cleaner to add example usage to the top of the script along with these details vs having to read through the code to understand the script inputs.

File: tests/e2e/faultinjection/duplicate_node_id.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+var _ = ginkgo.Describe("Duplicate node handling", func() {
+	require := require.New(ginkgo.GinkgoT())
+
+	ginkgo.It("should ensure that a given Node ID (i.e. staking keypair) can be used at most once on a network", func() {
+		network := e2e.Env.GetNetwork()
+		nodes := network.GetNodes()
+
+		ginkgo.By("creating new node")
+		node1 := e2e.AddTemporaryNode(network, testnet.FlagsMap{})
+		e2e.WaitForHealthy(node1)
```
Comment (hexfusion) [inline tests/e2e/faultinjection/duplicate_node_id.go:29]: Just so I understand better why here is `WaitForHealthy` takes a node but below it is a method on node. Just generally do we need both ways to do this? Regardless they should be the same through the test. I could be missing something though.

File: tests/fixture/testnet/local/network.go [hunk 5]
```diff
@@ -632,3 +644,48 @@ func (ln *LocalNetwork) ReadAll() error {
 	}
 	return ln.ReadNodes()
 }
+
+func (ln *LocalNetwork) AddLocalNode(w io.Writer, node *LocalNode) (*LocalNode, error) {
+	// Assume network configuration has been written to disk and is current in memory
+
+	if node == nil {
+		node = NewLocalNode("")
```
Comment (hexfusion) [inline tests/fixture/testnet/local/network.go:652]: can we add a comment on what is happening here. is this a node with no name or default or?

File: tests/fixture/testnet/local/node.go [hunk 6]
```diff
@@ -186,13 +186,24 @@ func (n *LocalNode) Start(w io.Writer, defaultExecPath string) error {
 		return err
 	}
 
+	// A temporary node directory won't match its node name.
+	isTemporaryNode := filepath.Base(n.GetDataDir()) != n.NodeID.String()
+	var nodeDescription string
+	if isTemporaryNode {
+		// Qualify the description for a temporary node with its path
+		// since it won't be stored in the network's directory.
+		nodeDescription = fmt.Sprintf("temporary node %q @ %s", n.NodeID, n.GetDataDir())
```
Comment (hexfusion) [inline tests/fixture/testnet/local/node.go:195]:
```suggestion
		nodeDescription = fmt.Sprintf("temporary node: %q path: %s", n.NodeID, n.GetDataDir())
```

File: tests/e2e/e2e.go (hunk 1 above)
Comment (hexfusion) [reply tests/e2e/e2e.go:37]: Ok so double observed upper bound, makes sense.

File: scripts/tests.e2e.sh (hunk 3 above)
Comment (hexfusion) [reply scripts/tests.e2e.sh:54]: Sorry if that was not clear I am not questioning the check I am questioning the docs for the ginkgo args. It feels like it should be at the top of the doc maybe above or below here[1]. Just so user doesn't need to sift through the code to understand the params. Jjust an observation.

https://github.com/ava-labs/avalanchego/blob/6e6b19a3909e9aba8e83b3882a5377ef2c9b16ec/scripts/tests.e2e.sh#L5-L7

File: tests/e2e/faultinjection/duplicate_node_id.go (hunk 4 above)
Comment (hexfusion) [reply tests/e2e/faultinjection/duplicate_node_id.go:29]: OK I think it is strange that we have two methods with the same name and different functionality and signatures. IMO `WaitForHealthy` should always take a ctx and then allow the test writer the choice of timeout. If a method infers a set timeout that should be something else?

File: tests/fixture/testnet/local/node.go [hunk 7]
```diff
@@ -308,7 +319,28 @@ func (n *LocalNode) IsHealthy(ctx context.Context) (bool, error) {
 		}
 	}
 	// Assume all other errors are not recoverable
-	return false, err
+	return false, fmt.Errorf("failed to query node health: %w", err)
+}
+
+func (n *LocalNode) WaitForHealthy(ctx context.Context) error {
```
Comment (hexfusion) [inline tests/fixture/testnet/local/node.go:325]: can we get some docs on this method ?

**Review by hexfusion — APPROVED** [review]

File: tests/fixture/testnet/local/node.go [hunk 8]
```diff
@@ -308,7 +320,32 @@ func (n *LocalNode) IsHealthy(ctx context.Context) (bool, error) {
 		}
 	}
 	// Assume all other errors are not recoverable
-	return false, err
+	return false, fmt.Errorf("failed to query node health: %w", err)
+}
+
+// WaitForHealthy blocks until IsHealthy returns true or an error (including context timeout) is observed.
+func (n *LocalNode) WaitForHealthy(ctx context.Context) error {
```
Comment (StephenButtolph) [inline tests/fixture/testnet/local/node.go:327]: Can we remove this function from the `Node` interface and instead make this a function that takes in any implementation of the `Node` interface?

File: tests/fixture/testnet/local/network.go [hunk 9]
```diff
@@ -632,3 +649,65 @@ func (ln *LocalNetwork) ReadAll() error {
...
+	node.SetNetworkingConfigDefaults(httpPort, stakingPort, bootstrapIDs, bootstrapIPs)
+
+	if err := node.WriteConfig(); err != nil {
+		return nil, err
+	}
+	if err := node.Start(w, ln.ExecPath); err != nil {
+		return nil, err
+	}
+
+	return node, nil
```
Comment (StephenButtolph) [inline tests/fixture/testnet/local/network.go:712]:
```suggestion
	return node, node.Start(w, ln.ExecPath)
```

File: tests/fixture/testnet/local/network.go [hunk 10]
```diff
@@ -632,3 +649,65 @@ func (ln *LocalNetwork) ReadAll() error {
...
+	// Use dynamic port allocation.
+	var httpPort uint16 = 0
+	var stakingPort uint16 = 0
+
+	// Collect staking addresses of running nodes for use in bootstraping the new node
+	if err := ln.ReadNodes(); err != nil {
+		return nil, fmt.Errorf("failed to read local network nodes: %w", err)
+	}
+	bootstrapIPs := make([]string, 0, len(ln.Nodes))
+	bootstrapIDs := make([]string, 0, len(ln.Nodes))
```
Comment (StephenButtolph) [inline tests/fixture/testnet/local/network.go:689]:
```suggestion
	// Collect staking addresses of running nodes for use in bootstraping the new node
	if err := ln.ReadNodes(); err != nil {
		return nil, fmt.Errorf("failed to read local network nodes: %w", err)
	}
	
	var (
		// Use dynamic port allocation.
		httpPort    uint16 = 0
		stakingPort uint16 = 0

		bootstrapIPs = make([]string, 0, len(ln.Nodes))
		bootstrapIDs = make([]string, 0, len(ln.Nodes))
	)
```

File: tests/e2e/faultinjection/duplicate_node_id.go [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+// See the file LICENSE for licensing terms.
+
+package faultinjection
+
+import (
+	"context"
+	"fmt"
+
+	ginkgo "github.com/onsi/ginkgo/v2"
+	"github.com/stretchr/testify/require"
```
Comment (StephenButtolph) [inline tests/e2e/faultinjection/duplicate_node_id.go:11]:
```suggestion
	ginkgo "github.com/onsi/ginkgo/v2"

	"github.com/stretchr/testify/require"
```

**Review by StephenButtolph — APPROVED** [review]
super nit-y, but after these lgtm

File: tests/fixture/testnet/local/network.go (hunk 9 above)
Comment (StephenButtolph) [reply tests/fixture/testnet/local/network.go:716]: We certainly do not follow that in avalanchego. AvalancheGo expects the caller to verify that the returned error is nil. If it is non-nil, there are no assumptions made about the other returned values.

**Pull request merged**